Mathlib/Algebra/Lie/Weights/Basic.lean

```diff
@@ -184,6 +185,48 @@ lemma weightSpace_le_weightSpaceOf (x : L) (χ : L → R) :
     weightSpace M χ ≤ weightSpaceOf M (χ x) x :=
   iInf_le _ x
 
+variable (R L) in
+/-- A weight of a Lie module is a map `L → R` such that the corresponding weight space is is
+non-trivial. -/
+structure Weight where
+  /-- The family of eigenvalues corresponding to a weight. -/
+  toFun : L → R
+  weightSpace_ne_bot : weightSpace M toFun ≠ ⊥
+
+namespace Weight
+
+instance instFunLike :
+    FunLike (Weight R L M) L R where
+  coe χ := χ.1
+  coe_injective' χ₁ χ₂ h := by cases χ₁; cases χ₂; simp_all
+
+@[simp] lemma coe_weight_mk (χ : L → R) (h) :
+    (↑(⟨χ, h⟩ : Weight R L M) : L → R) = χ :=
+  rfl
+
+variable {M}
+
+/-- The proposition that a weight of a Lie module is zero.
+
+We make this definition because we cannot define a `Zero (Weight R L M)` instance since the weight
+space of the zero function can be trivial. -/
+def IsZero (χ : Weight R L M) := (χ : L → R) = 0
+
+@[simp] lemma coe_eq_zero_iff (χ : Weight R L M) : (χ : L → R) = 0 ↔ χ.IsZero := Iff.rfl
+
+/-- The proposition that a weight of a Lie module is non-zero. -/
+abbrev IsNonZero (χ : Weight R L M) := ¬ IsZero (χ : Weight R L M)
+
+variable (R L M) in
+/-- The set of weights is equivalent to a subtype. -/
+def equivSetOf : Weight R L M ≃ {χ : L → R | weightSpace M χ ≠ ⊥} where
+  toFun w := ⟨w.1, w.2⟩
+  invFun w := ⟨w.1, w.2⟩
+  left_inv w := by simp
+  right_inv w := by simp
```

Comment:
Would it work to make this the actual definition? Or are there downsides to that?

Reply:
Dot notation fails and I have to manually author lemmas for named fields if I just use a `Subtype`. Possibly (though I'm not sure) I also encountered undesirable `simp` lemmas when I experimented with `Subtype` instead of `structure`.

So I think this is the right choice but I'm happy to investigate further if you wish.

Reply:
No need to investigate further, I'm convinced (-;